sdk/storage/Azure.Storage.Files.DataLake/src/DataLakeDirectoryClient.cs

```diff
@@ -2187,5 +2188,95 @@ public virtual DataLakeDirectoryClient GetSubDirectoryClient(string subdirectory
             }
         }
         #endregion Delete Sub Directory
+
+        #region Get Paths
+        /// <summary>
+        /// The <see cref="GetPaths"/> operation returns an async sequence
+        /// of paths in this directory.  Enumerating the paths may make
+        /// multiple requests to the service while fetching all the values.
+        ///
+        /// For more information, see
+        /// <see href="https://docs.microsoft.com/rest/api/storageservices/datalakestoragegen2/path/list">
+        /// List Path(s)</see>.
+        /// </summary>
+        /// <param name="recursive">
+        /// If "true", all paths are listed; otherwise, only paths at the root of the directory are listed.
+        /// </param>
+        /// <param name="userPrincipalName">
+        /// Optional. Valid only when Hierarchical Namespace is enabled for the account. If
+        /// "true", the user identity values returned in the owner and group fields of each list
+        /// entry will be transformed from Azure Active Directory Object IDs to User Principal
+        /// Names. If "false", the values will be returned as Azure Active Directory Object IDs.
+        /// The default value is false. Note that group and application Object IDs are not translated
+        /// because they do not have unique friendly names.
+        /// </param>
+        /// <param name="cancellationToken">
+        /// Optional <see cref="CancellationToken"/> to propagate
+        /// notifications that the operation should be cancelled.
+        /// </param>
+        /// <returns>
+        /// An <see cref="Pageable{PathItem}"/>
+        /// describing the paths in the directory.
+        /// </returns>
+        /// <remarks>
+        /// A <see cref="RequestFailedException"/> will be thrown if
+        /// a failure occurs.
+        /// </remarks>
+        public virtual Pageable<PathItem> GetPaths(
+            bool recursive = default,
+            bool userPrincipalName = default,
```

Comment:
should this be options bag?

Reply:
I think this is the way it is for `DataLakeFileSystemClient.GetPaths()` too. I do like the idea of an options bag though, but if we do it here, we should go back and do it for the FileSystemClient too and hide the old one.

Reply:
I was also thinking about this when I did the implementation, I wanted to see what y'all thought.  I'll switch both to options bags.

Reply:
Fixed.

Reply:
Do we think this will add a lot of parameters in the future?  You currently have 2 and the informal guideline is roughly 6.

Reply:
It's possible it could grow, but I don't know that there are any specific plans.  I added this based on @kasobol-msft and @amnguye's comments.

Reply:
I'd follow [this](https://www.matheus.ro/2018/01/29/clean-code-avoid-many-arguments-functions/) if there's no formal guidance. Let's not forget cancellationToken when we count parameters.

Reply:
That link offers a highly opinionated view that clashes with the rest of our (different highly opinionated :smile:) style.  In general for questions like this we should get them clarified in the guidelines rather than seek elsewhere so Storage libraries look and feel the same as other Azure services.  I checked https://azure.github.io/azure-sdk/dotnet_introduction.html#dotnet-parameters and it does now list 6 parameters:

> Simple methods are methods that take up to six parameters, with most of the parameters being simple BCL primitives.  ... Simple methods should follow standard .NET Design Guidelines for parameter list and overload design.

Unless we expect this to more than double, I'd suggest using parameters for now.

Reply:
Surprisingly the https://azure.github.io/azure-sdk/dotnet_introduction.html#dotnet-parameters gives an example of "simple method" that has 3 parameters (2 primitves + cancellation token). In general I think 6 is too much and I think the sdk guide should trim that to 3 (6 parameter method doesn't look good imo). 
Anyway in that case since we 2 params + token let's follow the guideline then and go back to original design.